// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// =============================================================================
syntax = "proto3";

package tensorflow.data_validation;

import "tensorflow/core/example/example.proto";

// Skew information for a particular feature.
message FeatureSkew {
  // Name of the feature.
  string feature_name = 1;

  // All of the counts are with respect to the set of tensorflow examples
  // which are found in **BOTH** the baseline and testing data.
  // Number of examples in baseline and test where this feature is present.
  uint64 base_count = 2;
  uint64 test_count = 3;

  // Number of examples where feature is present in both baseline and
  // test data, and the feature value matches exactly.
  uint64 match_count = 4;

  // Number of times feature is present for baseline or test only.
  uint64 base_only = 5;
  uint64 test_only = 6;

  // Total number of examples where the feature is present in both
  // baseline and test but the feature value does not match.
  uint64 mismatch_count = 7;

  // diff_count == base_only + test_only + mismatch_count
  uint64 diff_count = 8;
}

// Message to store a pair of tensorflow examples which exhibit some skew.
message SkewPair {
  // Example as it appears in baseline data.
  tensorflow.Example base = 1;

  // Example from test data with the same identifier as that of the baseline.
  tensorflow.Example test = 2;

  // Features that appear in the baseline example but not in test.
  repeated string base_only_features = 3;

  // Features that appear in the test example but not in baseline.
  repeated string test_only_features = 4;

  // Features that appear in both baseline and test example, and their
  // feature values are matched.
  repeated string matched_features = 5;

  // Features that appear in both baseline and test example, and their
  // feature values do not match.
  repeated string mismatched_features = 6;
}
